hazelcast/test/src/compact_test.cpp

```diff
@@ -506,6 +538,160 @@ TEST_F(CompactSerializationTest, test_rabin_fingerprint_consistent_with_server)
+        object.long_ = reader.read_int64("long");
+        object.float_ = reader.read_float32("float");
+        object.double_ = reader.read_float64("double");
+        return object;
+    }
+
+    static std::string type_name() { return "primitive_object"; }
+};
+
+template<>
+struct hz_serializer<compact::test::nullable_primitive_object>
+  : public compact_serializer
+{
+    static void write(const compact::test::nullable_primitive_object& object,
+                      compact_writer& writer)
+    {
+        writer.write_nullable_boolean("boolean", object.nullableBoolean);
+        writer.write_nullable_int8("byte", object.nullableByte);
+        writer.write_nullable_int16("short", object.nullableShort);
+        writer.write_nullable_int32("int", object.nullableInt);
+        writer.write_nullable_int64("long", object.nullableLong);
+        writer.write_nullable_float32("float", object.nullableFloat);
+        writer.write_nullable_float64("double", object.nullableDouble);
+    }
+
+    static compact::test::nullable_primitive_object read(compact_reader& reader)
+    {
+        compact::test::nullable_primitive_object object;
+        object.nullableBoolean = reader.read_nullable_boolean("boolean");
+        object.nullableByte = reader.read_nullable_int8("byte");
+        object.nullableShort = reader.read_nullable_int16("short");
+        object.nullableInt = reader.read_nullable_int32("int");
+        object.nullableLong = reader.read_nullable_int64("long");
+        object.nullableFloat = reader.read_nullable_float32("float");
+        object.nullableDouble = reader.read_nullable_float64("double");
+        return object;
+    }
+
+    // typename is same as the primitive_object on purpose.
+    // This is to simulate two different applications implementing the same
+    // class with different serializers.
+    static std::string type_name() { return "primitive_object"; }
+};
+} // namespace serialization
+namespace compact {
+namespace test {
+class CompactNullablePrimitiveInteroperabilityTest : public ::testing::Test
+{};
+TEST_F(CompactNullablePrimitiveInteroperabilityTest,
+       testWritePrimitiveReadNullable)
+{
+    primitive_object expected{ true, 2, 4, 8, 4444, 8321.321F, 41231.32 };
+    serialization_config config;
+    SerializationService ss(config);
+
+    const data& data = ss.to_data(expected);
+    auto actual = ss.to_object<nullable_primitive_object>(data).value();
+    ASSERT_EQ(expected.boolean_, actual.nullableBoolean.value());
+    ASSERT_EQ(expected.byte_, actual.nullableByte.value());
+    ASSERT_EQ(expected.short_, actual.nullableShort.value());
+    ASSERT_EQ(expected.int_, actual.nullableInt.value());
+    ASSERT_EQ(expected.long_, actual.nullableLong.value());
+    ASSERT_EQ(expected.float_, actual.nullableFloat.value());
+    ASSERT_EQ(expected.double_, actual.nullableDouble.value());
+}
+
+TEST_F(CompactNullablePrimitiveInteroperabilityTest,
+       testWriteNullableReadPrimitive)
```

Comment:
Maybe we can rename the test to testWriteNullableAsNonNullReadPrimitive

Reply:
Test names are copied from java. Since there is not much difference, I will not change it to keep it consistent with java

Reply:
Hmm I thought these are from python client. Ok then